// Controllers/DietController.js
var Diets 	= require('DB.Modules/Diets'),
	v 		= require('Validate-alpha'),
	layout 	= require('Layout');
/*
 * Diet Controller
 */
DietController = {
	all: function(){
		Diets.where('deleted IS NULL', 10, function(opt){
			opt.routeName = "diets-all";
			layout.LoadContent('diets/all', opt);
		});
	},
	getAdd: function(options){
		layout.LoadContent('diets/add', options);
		window.tinymce.init({
			selector:'textarea'
		});
	},
	postAdd: function(){
		// collect data
		var data = {
			name : doc.getElementById('dietName').value,
			details: doc.getElementById('dietText').value
		},

		// validate data
		val = {
			name: ['required', 'alpha', 'min:3', 'unique:diets,name'],
			details: ['required', 'min:10']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var options = {
					inputs: data,
					errors: errors
				};
				redirect('diets-add', options);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Added Failed",
					bodyContent: html,
					hide: function(){}
				});
			} else {
				// save data
				var obj = {
					name : data.name,
					details: data.details,
					created_at: Diets.currentDateTime(),
					updated_at: Diets.currentDateTime()
				};
				Diets.create(obj, function(newId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Diet Added",
						bodyContent: "<p>a new diet ("+ data.name +") added to the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	profile: function(options){
		Diets.find(options.id, function(rows){
			options.data = rows[0];
			layout.LoadContent('diets/profile', options);
		});
	},
	update: function(options){
		// collect data
		var data = {
			name : doc.getElementById('dietName').value,
			details: doc.getElementById('dietText').value
		},

		// validate data
		val = {
			name: ['required', 'min:3', 'unique:diets,name,'+ options.id],
			details: ['required', 'min:10']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var retrn = {
					inputs: data,
					errors: errors,
					id: options.id
				};
				redirect('diets-profile', retrn);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Updated Failed",
					bodyContent: html,
					hide: function(){}
				});
			} else {
				// save data
				var obj = {
					name : data.name,
					details: data.details,
					updated_at: Diets.currentDateTime()
				};
				Diets.update(options.id, obj, function(affectedId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Diet Updated",
						bodyContent: "<p>diet ("+ data.name +") Updated Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	remove: function(options){
		HCalert({
			type: "error", // success, updated, error
			headerText: "Deleteing Diet",
			bodyContent: "<p>are you sure you want to delete this Diet!!</p><p>you CAN'T undo this operation!!!</p>",
			hide: function(){},
			ok: function(){
				Diets.softDelete(options.id, function(affecedId){
					redirect('diets-all');
					HCalert({
						type: "success", // success, updated, error
						headerText: "Diet Deleted",
						bodyContent: "<p>diet deleted from the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	}
}
module.exports = DietController;
