// Controllers/GroupController.js
var Groups 	= require('DB.Modules/Groups'),
	v 		= require('Validate-alpha'),
	layout 	= require('Layout');
/*
 * Group Controller
 */
GroupController = {
	all: function(){
		Groups.where('deleted IS NULL', 10, function(opt){
			opt.routeName = "groups-all";
			layout.LoadContent('groups/all', opt);
		});
	},
	getAdd: function(options){
		layout.LoadContent('groups/add', options);
	},
	postAdd: function(){
		// collect data
		var data = {
			name : 		doc.getElementById('groupName').value,
			discount: 	doc.getElementById('groupDiscount').value,
			fixed_price:doc.getElementById('groupFixed').value
		},

		// validate data
		val = {
			name: 		['required', 'min:3', 'alpha', 'unique:groups,name'],
			discount: 	['numaric'],
			fixed_price:['numaric']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var options = {
					inputs: data,
					errors: errors
				};
				redirect('groups-add', options);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Added Failed",
					bodyContent: html,
					hide: function(){}
				});
			} else {
				// save data
				var obj = {
					name : 		data.name,
					discount: 	data.discount,
					fixed: 		data.fixed_price,
					created_at: Groups.currentDateTime(),
					updated_at: Groups.currentDateTime()
				};
				Groups.create(obj, function(newId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Group Added",
						bodyContent: "<p>a new Group ("+ data.name +") added to the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	profile: function(options){
		Groups.find(options.id, function(rows){
			options.data = rows[0];
			layout.LoadContent('groups/profile', options);
		});
	},
	update: function(options){
		// collect data
		var data = {
			name : 		doc.getElementById('groupName').value,
			discount: 	doc.getElementById('groupDiscount').value,
			fixed_price:doc.getElementById('groupFixed').value
		},

		// validate data
		val = {
			name: 		['required', 'unique:groups,name,'+ options.id, 'min:3', 'alpha'],
			discount: 	['numaric'],
			fixed_price:['numaric']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var retrn = {
					inputs: data,
					errors: errors,
					id: options.id
				};
				redirect('groups-profile', retrn);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Updated Failed",
					bodyContent: html,
					hide: function(){}
				});
				console.log(retrn);
			} else {
				// save data
				var obj = {
					name : data.name,
					discount: data.discount,
					fixed: data.fixed_price,
					updated_at: Groups.currentDateTime()
				};
				Groups.update(options.id, obj, function(affectedId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Group Updated",
						bodyContent: "<p>group ("+ data.name +") Updated Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	remove: function(options){
		HCalert({
			type: "error", // success, updated, error
			headerText: "Deleteing Group",
			bodyContent: "<p>are you sure you want to delete this Group!!</p><p>you CAN'T undo this operation!!!</p>",
			hide: function(){},
			ok: function(){
				Groups.softDelete(options.id, function(affecedId){
					redirect('groups-all');
					HCalert({
						type: "success", // success, updated, error
						headerText: "Group Deleted",
						bodyContent: "<p>group deleted from the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	}
}
module.exports = GroupController;
