// Controllers/PageController.js
/*
 * Static Pages Controller
 */
var layout = require('Layout');

var PageController = {
	home: function(options){
		if ( doc.getElementById('header') ) {
			layout.LoadPage('home');
		} else {
			layout.Load('inside', options);
			layout.LoadPage('home');
		};
	},
	about: function(){
		if ( doc.getElementById('inside-page') )
			layout.LoadPage('about');
		else {
			HCalert({
				type: "error",
				headerText: "Page Loaded Failed",
				bodyContent: "You are NOT sign in yet.",
				hide: function(){}
			});
		}
	},
	license: function(){
		if ( doc.getElementById('inside-page') )
			layout.LoadPage('license');
		else {
			HCalert({
				type: "error",
				headerText: "Page Loaded Failed",
				bodyContent: "You are NOT sign in yet.",
				hide: function(){}
			});
		}
	},
	help: function(){
		layout.LoadPage('help');
	},
	options: function(){
		layout.LoadPage('options');
	},
	patients: function(){
		layout.Columns();
		layout.LoadSidebar('patient');
		new require('Routes')('patients-all');
	},
	admin: function(){
		layout.Columns();
		layout.LoadSidebar('admin');
		new require('Routes')('employees-all');
	},
	accounts: function(){
		layout.Columns();
		layout.LoadSidebar('account');
		new require('Routes')('store-measurement');
	},
	reports: function(){
		layout.Columns();
		layout.LoadSidebar('report');
	}
}
module.exports = PageController;
