// Controllers/SessionController.js
var Sessions 	= require('DB.Modules/Sessions'),
	Categories 	= require('DB.Modules/Categories'),
	v 			= require('Validate-alpha'),
	layout 		= require('Layout');
/*
 * Group Controller
 */
SessionController = {
	all: function(){
		Sessions.where('deleted IS NULL', 10, function(opt){
			opt.routeName = "sessions-all";
			layout.LoadContent('sessions/all', opt);
		});
	},
	getAdd: function(options){
		Categories.all('', function(cats){
			options.cats = cats.rows;
			layout.LoadContent('sessions/add', options);
		});
	},
	postAdd: function(){
		// collect data
		var data = {
			catId : 	doc.getElementById('categoryId').value,
			name : 		doc.getElementById('sessionName').value,
			price : 	doc.getElementById('sessionPrice').value,
			machineIds :doc.getElementById('machinesIds').value,
		},

		// validate data
		val = {
			name: 		['required', 'min:3', 'alpha', 'unique:sessions,name'],
			price: 		['required', 'numaric'],
			machineIds: ['required']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var options = {
					inputs: data,
					errors: errors
				};
				redirect('Sessions-add', options);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Added Failed",
					bodyContent: html,
					hide: function(){}
				});
			} else {
				// save data
				var obj = {
					name : 		data.name,
					discount: 	data.discount,
					fixed: 		data.fixed_price,
					created_at: Sessions.currentDateTime(),
					updated_at: Sessions.currentDateTime()
				};
				Sessions.create(obj, function(newId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Group Added",
						bodyContent: "<p>a new Group ("+ data.name +") added to the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	profile: function(options){
		Sessions.find(options.id, function(rows){
			options.data = rows[0];
			layout.LoadContent('sessions/profile', options);
		});
	},
	update: function(options){
		// collect data
		var data = {
			name : 		doc.getElementById('groupName').value,
			discount: 	doc.getElementById('groupDiscount').value,
			fixed_price:doc.getElementById('groupFixed').value
		},

		// validate data
		val = {
			name: 		['required', 'unique:sessions,name,'+ options.id, 'min:3', 'alpha'],
			discount: 	['numaric'],
			fixed_price:['numaric']
		};
		v.make(data, val, function(errors){
			var keys = Object.keys(errors),
				length = keys.length;
			if (length) {
				var html = "";
				for (var i=0; i < length; i++) {
					html += "<h3>"+ keys[i] +"</h3>";
					html += "<ul>";
						for (var z=0; z < errors[keys[i]].length; z++) {
							html += "<li>"+ errors[keys[i]][z] +"</li>";
						};
					html += "</ul>";
				};
				var retrn = {
					inputs: data,
					errors: errors,
					id: options.id
				};
				redirect('Sessions-profile', retrn);
				HCalert({
					type: "error", // success, updated, error
					headerText: "Updated Failed",
					bodyContent: html,
					hide: function(){}
				});
				console.log(retrn);
			} else {
				// save data
				var obj = {
					name : data.name,
					discount: data.discount,
					fixed: data.fixed_price,
					updated_at: Sessions.currentDateTime()
				};
				Sessions.update(options.id, obj, function(affectedId){
					HCalert({
						type: "success", // success, updated, error
						headerText: "Group Updated",
						bodyContent: "<p>group ("+ data.name +") Updated Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	},
	remove: function(options){
		HCalert({
			type: "error", // success, updated, error
			headerText: "Deleteing Session",
			bodyContent: "<p>are you sure you want to delete this Session!!</p><p>you CAN'T undo this operation!!!</p>",
			hide: function(){},
			ok: function(){
				Sessions.softDelete(options.id, function(affecedId){
					redirect('sessions-all');
					HCalert({
						type: "success", // success, updated, error
						headerText: "Session Deleted",
						bodyContent: "<p>session deleted from the system Successfully.</p>",
						hide: function(){}
					});
				});
			}
		});
	}
}
module.exports = SessionController;
